// ts2dart/build/lib/base.js
/// <reference path='../node_modules/typescript/bin/typescript.d.ts' />
var ts = require('typescript');
function ident(n) {
    if (n.kind === 65 /* Identifier */)
        return n.text;
    if (n.kind === 126 /* QualifiedName */) {
        var qname = n;
        var leftName = ident(qname.left);
        if (leftName)
            return leftName + '.' + ident(qname.right);
    }
    return null;
}
exports.ident = ident;
var TranspilerStep = (function () {
    function TranspilerStep(transpiler) {
        this.transpiler = transpiler;
    }
    TranspilerStep.prototype.visit = function (n) { this.transpiler.visit(n); };
    TranspilerStep.prototype.emit = function (s) { this.transpiler.emit(s); };
    TranspilerStep.prototype.emitNoSpace = function (s) { this.transpiler.emitNoSpace(s); };
    TranspilerStep.prototype.reportError = function (n, message) { this.transpiler.reportError(n, message); };
    TranspilerStep.prototype.visitNode = function (n) { throw Error('not implemented'); };
    TranspilerStep.prototype.visitEach = function (nodes) {
        var _this = this;
        nodes.forEach(function (n) { return _this.visit(n); });
    };
    TranspilerStep.prototype.visitEachIfPresent = function (nodes) {
        if (nodes)
            this.visitEach(nodes);
    };
    TranspilerStep.prototype.visitList = function (nodes, separator) {
        if (separator === void 0) { separator = ','; }
        for (var i = 0; i < nodes.length; i++) {
            this.visit(nodes[i]);
            if (i < nodes.length - 1)
                this.emit(separator);
        }
    };
    TranspilerStep.prototype.getAncestor = function (n, kind) {
        for (var parent = n; parent; parent = parent.parent) {
            if (parent.kind === kind)
                return parent;
        }
        return null;
    };
    TranspilerStep.prototype.hasAncestor = function (n, kind) { return !!this.getAncestor(n, kind); };
    TranspilerStep.prototype.hasAnnotation = function (decorators, name) {
        if (!decorators)
            return false;
        return decorators.some(function (d) {
            var decName = ident(d.expression);
            if (decName === name)
                return true;
            if (d.expression.kind !== 157 /* CallExpression */)
                return false;
            var callExpr = d.expression;
            decName = ident(callExpr.expression);
            return decName === name;
        });
    };
    TranspilerStep.prototype.hasFlag = function (n, flag) {
        return n && (n.flags & flag) !== 0 || false;
    };
    TranspilerStep.prototype.isConst = function (decl) {
        var _this = this;
        return this.hasAnnotation(decl.decorators, 'CONST') ||
            decl.members.some(function (m) {
                if (m.kind !== 135 /* Constructor */)
                    return false;
                return _this.hasAnnotation(m.decorators, 'CONST');
            });
    };
    // TODO(martinprobst): This belongs to module.ts, refactor.
    TranspilerStep.prototype.getLibraryName = function () { return this.transpiler.getLibraryName(); };
    TranspilerStep.prototype.visitTypeName = function (typeName) {
        if (typeName.kind !== 65 /* Identifier */) {
            this.visit(typeName);
            return;
        }
        var identifier = ident(typeName);
        if (TranspilerStep.DART_TYPES.hasOwnProperty(identifier)) {
            identifier = TranspilerStep.DART_TYPES[identifier];
        }
        this.emit(identifier);
    };
    TranspilerStep.prototype.maybeVisitTypeArguments = function (n) {
        if (n.typeArguments) {
            this.emit('<');
            this.visitList(n.typeArguments);
            this.emit('>');
        }
    };
    TranspilerStep.DART_TYPES = {
        'Promise': 'Future',
        'Observable': 'Stream',
        'ObservableController': 'StreamController',
        'Date': 'DateTime',
        'StringMap': 'Map'
    };
    return TranspilerStep;
})();
exports.TranspilerStep = TranspilerStep;
